sodetlib/operations/uxm_setup.py

```diff
@@ -363,6 +363,162 @@ def setup_tune(S, cfg, bands, show_plots=False, update_cfg=True):
     return True, summary
 
 
+def _find_fixed_tone_freq(S, band, ntone, min_spacing, band_start=-250, band_end=250):
+    """Place fixed tones in the largest gaps between resonators."""
+
+    # get resonator and channel frequencies
+    res_freq = np.sort([res["freq"] for res in S.freq_resp[band]["resonances"].values()])
+
+    # only consider central part of the band
+    band_center = S.get_band_center_mhz(band)
+    freq_low = band_center + band_start
+    freq_high = band_center + band_end
+    res_freq = res_freq[(res_freq > freq_low) & (res_freq < freq_high)]
+
+    # gaps between resonators
+    gaps = np.diff(res_freq)
+
+    # check there are enough for the number of tones
+    num_gaps = (gaps >= min_spacing).sum()
+    if num_gaps < ntone:
+        S.log(
+            f"Only {num_gaps} gaps greater than {min_spacing} MHz were found in band {band}. "
+            "That number of fixed tones will be set."
+        )
+        ntone = num_gaps
+
+    # return the central frequencies in selected gaps
+    gap_ind = np.argsort(gaps)[:-(ntone + 1):-1]  # ntone largest gaps
+    return 0.5 * (res_freq[gap_ind + 1] + res_freq[gap_ind])
+
+
+@sdl.set_action()
+def setup_fixed_tones(
+    S, cfg, fixed_tones_per_band=8, min_gap=10, tone_power=10, update_cfg=True
+    ):
+    """
+    Identify gaps between resonators and attempt to place a number of fixed tones there.
+    These will be saved for every band in the device config. A tune should be available
+    before running this action.
+
+    Args
+    ----
+    S : SmurfControl
+        Pysmurf instance
+    cfg : DetConfig
+        DetConfig instance
+    fixed_tones_per_band : int
+        The number of tones to place in each band. Defaults to 8.
+    min_gap : float
+        The minimum gap size in order to place a fixed tone, in MHz. Defaults to 10.
+    tone_power : int
+        The tone power to use for the fixed tones. Defaults to 10.
+    update_cfg : bool
+        Whether to save the fixed tone configuration to the device config. Defaults to True.
+
+    Returns
+    -------
+    fixed_tones : dict
+        Fixed tones configuration, as saved to device config.
+    """
+
+    bands = cfg.dev.exp['active_bands']
```

Comment:
Can you accept a `bands` arg of type Optional[List] which defaults to using active_bands?